### DISCUSSION TITLE
Revit api. Retrieve all parameters and categories

### DISCUSSION THREAD

**Discussion by Nice3point** [jeremytammik/RevitLookup#183] [Show and tell]
## Summary

I recently came across a problem from my business partner who wanted to get all the built-in Revit parameters to extract metadata such as data type, units, storage type, etc. As we know, this is not possible using Autodesk's public API. 

Similar situation with categories, we can't get all the built-in categories, the only available way is to get them from the document settings, but it contains a very truncated list

![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/faae850b-818d-463d-af52-bcd832fff416)

## Solution

We can use private code using reflection and pointers, we only need a document. 
The point of the method is not to get them but create it.

Example code to get all parameters:

```c#
public static List<Parameter> GetBuiltinParameters(Document document)
{
    const BindingFlags bindingFlags = BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly;
    
    var documentType = typeof(Document);
    var parameterType = typeof(Parameter);
    var assembly = Assembly.GetAssembly(parameterType);
    var aDocumentType = assembly.GetType("ADocument");
    var elementIdType = assembly.GetType("ElementId");
    var elementIdIdType = elementIdType.GetField("<alignment member>", bindingFlags)!;
    var getADocumentType = documentType.GetMethod("getADocument", bindingFlags)!;
    var parameterCtorType = parameterType.GetConstructor(bindingFlags, null, new[] {aDocumentType.MakePointerType(), elementIdType.MakePointerType()}, null)!;

    var builtinParameters = Enum.GetValues(typeof(BuiltInParameter));
    var parameters = new List<Parameter>(builtinParameters.Length);
    foreach (BuiltInParameter builtinParameter in builtinParameters)
    {
        var elementId = Activator.CreateInstance(elementIdType);
        elementIdIdType.SetValue(elementId, builtinParameter);

        var handle = GCHandle.Alloc(elementId);
        var elementIdPointer = GCHandle.ToIntPtr(handle);
        Marshal.StructureToPtr(elementId, elementIdPointer, true);

        var parameter = (Parameter) parameterCtorType.Invoke(new[] {getADocumentType.Invoke(document, null), elementIdPointer});
        parameters.Add(parameter);
        handle.Free();
    }

    return parameters;
}
```

Example code to get all categories:

```c#
public static List<Category> GetBuiltinCategories(Document document)
{
    const BindingFlags bindingFlags = BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly;

    var documentType = typeof(Document);
    var categoryType = typeof(Category);
    var assembly = Assembly.GetAssembly(categoryType);
    var aDocumentType = assembly.GetType("ADocument");
    var elementIdType = assembly.GetType("ElementId");
    var elementIdIdType = elementIdType.GetField("<alignment member>", bindingFlags)!;
    var getADocumentType = documentType.GetMethod("getADocument", bindingFlags)!;
    var categoryCtorType = categoryType.GetConstructor(bindingFlags, null, new[] {aDocumentType.MakePointerType(), elementIdType.MakePointerType()}, null)!;

    var builtInCategories = Enum.GetValues(typeof(BuiltInCategory));
    var categories = new List<Category>(builtInCategories.Length);
    foreach (BuiltInCategory builtInCategory in builtInCategories)
    {
        var elementId = Activator.CreateInstance(elementIdType);
        elementIdIdType.SetValue(elementId, builtInCategory);

        var handle = GCHandle.Alloc(elementId);
        var elementIdPointer = GCHandle.ToIntPtr(handle);
        Marshal.StructureToPtr(elementId, elementIdPointer, true);

        var category = (Category) categoryCtorType.Invoke(new[] {getADocumentType.Invoke(document, null), elementIdPointer});
        categories.Add(category);
        handle.Free();
    }

    return categories;
}
```

## Result

As a result, we have created all the parameters and all the categories of the entire Enum

Parameters:

![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/2fd8e4a6-cdec-4626-be10-4fc42e9023ac)

Categories:

![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/f9f8bdfe-fe80-4440-8092-9efab1868d0b)

## Limitations

Created parameters have no binding to the element, and consequently have no value, only metadata

**Comment by chuongmep** [comment]
@Nice3point it is awesome !

**Comment by chuongmep** [comment]
Hi @Nice3point , it look like not working in Revit 2025 ?

**Reply by Nice3point** [reply]
@chuongmep should work, RevitLookup gets the categories/parameters this way
![image](https://github.com/user-attachments/assets/82e7ca23-69b3-4ff0-8f97-9eff9b4d818c)
![image](https://github.com/user-attachments/assets/98281d53-5c35-4ab9-89f3-431fa07aecee)

**Comment by chuongmep** [comment]
I can confirm it working awesome, thank @Nice3point 
![image](https://github.com/user-attachments/assets/ddb9e6bf-9dca-487e-877d-3007d43cff1f)

**Reply by chuongmep** [reply]
I have some issue with duplidate parameter and I want to share for any have same issue : 
```cs
 public static List<Autodesk.Revit.DB.Parameter> GetBuiltinParameters(Document document)
    {
        const BindingFlags bindingFlags = BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly;

        var documentType = typeof(Document);
        var parameterType = typeof(Autodesk.Revit.DB.Parameter);
        var assembly = Assembly.GetAssembly(parameterType);
        var aDocumentType = assembly.GetType("ADocument");
        var elementIdType = assembly.GetType("ElementId");
        var elementIdIdType = elementIdType.GetField("<alignment member>", bindingFlags)!;
        var getADocumentType = documentType.GetMethod("getADocument", bindingFlags)!;
        var parameterCtorType = parameterType.GetConstructor(bindingFlags, null,
            new[] { aDocumentType.MakePointerType(), elementIdType.MakePointerType() }, null)!;

        var builtinParameters = Enum.GetValues(typeof(BuiltInParameter));
        var parameters = new List<Autodesk.Revit.DB.Parameter>(builtinParameters.Length);
        foreach (BuiltInParameter builtinParameter in builtinParameters)
        {
            var elementId = Activator.CreateInstance(elementIdType);
            elementIdIdType.SetValue(elementId, builtinParameter);

            var handle = GCHandle.Alloc(elementId);
            var elementIdPointer = GCHandle.ToIntPtr(handle);
            Marshal.StructureToPtr(elementId, elementIdPointer, true);

            var parameter = (Autodesk.Revit.DB.Parameter)parameterCtorType.Invoke(new[]
                { getADocumentType.Invoke(document, null), elementIdPointer });
            parameters.Add(parameter);
            handle.Free();
        }
        // unique by id and name 
        var uniqueParameters = parameters.GroupBy(p => new { p.Id, p.Definition?.Name })
            .Select(g => g.First())
            .ToList();
        return uniqueParameters;
    }
```

**Comment by jeremytammik** [comment]
just to clarify: these are the unique parameter *names*; the parameters themselves are not unique, of course, or, rather, several different parameters may share the identical display name.

**Reply by Nice3point** [reply]
Exactly! So removing duplicates by name may not be appropriate for all cases

**Reply by chuongmep** [reply]
I tried with code but look like it is double with  identical display name as well 
![image](https://github.com/user-attachments/assets/bafc8de3-9fdc-44d1-addb-0206ddb0b7a8)

**Reply by Nice3point** [reply]
![image](https://github.com/user-attachments/assets/81e7a2e9-b48e-4373-ac05-f531e02eb043)

this is an API issue, creating a parameter by identical IDs

![image](https://github.com/user-attachments/assets/88992257-401f-40b1-9d34-5998538e508b)

**Comment by jeremytammik** [comment]
this is a fundamental parameter design issue. the ongoing project to convert the revit parameter definitions to some kind of globally unique repository that can be shared across all BIMs world-wide should resolve this. i searched on the roadmap and did not find exactly what i was expecting, but this thing here comes close: https://portal.productboard.com/mhf8vce9ubfupbhkfomayvi1/c/61-parameters-classification-taxonomy

**Reply by Nice3point** [reply]
also, for user parameters https://portal.productboard.com/mhf8vce9ubfupbhkfomayvi1/c/14-parameter-duplication-management

**Comment by Nice3point** [comment]
![image](https://github.com/user-attachments/assets/0baf8fca-f842-4386-91b0-67f400d9949e)

I've handled this issue, RevitLookup will now display the name as in Enum. However, when going to "**explore members**" it will still be a one parameter as designed in the API

@chuongmep 

The fix requires an additional call to GetNames, since in C# GetValues casts them to the same name
```c#
        var parameters = Enum.GetValues<BuiltInParameter>();
        var parameterNames = Enum.GetNames<BuiltInParameter>();
        var result = new List<UnitInfo>(parameters.Length);
        for (var i = 0; i < parameters.Length; i++)
        {
            ...
            var parameter = parameters[i];
            var name = parameterNames[i],
        }
```

**Reply by chuongmep** [reply]
That awesome, now I can see more clearly, Thanks @Nice3point
